### PR TITLE
Fixes #1866 | Adds ClayNavigationBar component

### LINKED ISSUE #1866: Create NavigationBar component

### PR DESCRIPTION
**Pull request by diegonvs**

### CODE REVIEW

File: packages/clay-navigation-bar/src/ClayNavigationBar.tsx [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	className,
+	id,
+	inverted = false,
+	items,
+	onItemClicked,
+	spritemap,
+	...otherProps
+}) => {
+	const [visible, setVisible] = React.useState(false);
+	const [transitioning, setTransitioning] = React.useState(false);
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/ClayNavigationBar.tsx]: Is it possible to extract some of this transitioning logic and height logic to its own hook? There is a lot of functions being created below here that are a bit hard to follow.

File: packages/clay-navigation-bar/src/ClayNavigationBar.tsx [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+					ref={contentRef}
+				>
+					<div className="container-fluid container-fluid-max-xl">
+						<ul className="navbar-nav">
+							{items.map(({active, href, label}, index) => (
+								<li
+									className={classNames('nav-item', {
+										active,
+									})}
+									data-nav-item-index={index}
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/ClayNavigationBar.tsx]: What is this for?

File: packages/clay-navigation-bar/src/ClayNavigationBar.tsx [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+import ClayButton from '@clayui/button';
+import ClayIcon from '@clayui/icon';
+import ClayLink from '@clayui/link';
+
+interface Props extends React.HTMLAttributes<HTMLDivElement> {
+	/**
+	 * This property is used to pass a MouseEvent from a NavItem to a
+	 * function when this element is clicked.
+	 */
+	onItemClicked?: (
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/ClayNavigationBar.tsx]: sort props

File: packages/clay-navigation-bar/src/ClayNavigationBar.tsx (hunk 1 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/ClayNavigationBar.tsx]: Yes! I'm taking a look!

File: packages/clay-navigation-bar/src/ClayNavigationBar.tsx (hunk 2 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/ClayNavigationBar.tsx]: Oh, I forgot. Isn't needed.

File: packages/clay-navigation-bar/src/ClayNavigationBar.tsx (hunk 3 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/ClayNavigationBar.tsx]: Done

File: packages/clay-navigation-bar/src/ClayNavigationBar.tsx (hunk 1 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/ClayNavigationBar.tsx]: I'm taking a look on that :)

File: packages/clay-navigation-bar/src/ClayNavigationBar.tsx [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+				'navbar-collapse-absolute',
+				'navbar-expand-md',
+				'navbar-underline',
+				'navigation-bar',
+				{
+					'navigation-bar-light': !inverted,
+					'navigation-bar-secondary': inverted,
+				}
+			)}
+			id={id}
```
[the remark this answers is not in the document]
Comment (diegonvs) [reply packages/clay-navigation-bar/src/ClayNavigationBar.tsx]: Makes sense!

File: packages/clay-navigation-bar/package.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+	},
+	"main": "lib/ClayNavigationBar.js",
+	"esnext:main": "src/ClayNavigationBar.js",
+	"jsnext:main": "src/ClayNavigationBar.js",
+	"files": ["lib", "src"],
+	"scripts": {
+		"build": "cross-env NODE_ENV=production babel src --root-mode upward --out-dir lib --extensions .ts,.tsx",
+		"build:types": "cross-env NODE_ENV=production tsc --project ./tsconfig.declarations.json",
+		"prepublishOnly": "npm run build && npm run build:types",
+		"start": "webpack-dev-server --mode development",
```
[the remark this answers is not in the document]
Comment (diegonvs) [reply packages/clay-navigation-bar/package.json]: Done, I was using an older version of our generator

File: packages/clay-navigation-bar/package.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+	"version": "3.0.0",
+	"description": "ClayNavigationBar component",
+	"license": "BSD-3-Clause",
+	"repository": "https://github.com/liferay/clay/tree/master/packages/clay-navigation-bar",
+	"engines": {
+		"node": ">=0.12.0",
+		"npm": ">=3.0.0"
+	},
+	"main": "lib/ClayNavigationBar.js",
+	"esnext:main": "src/ClayNavigationBar.js",
```
[the remark this answers is not in the document]
Comment (diegonvs) [reply packages/clay-navigation-bar/package.json]: Done!

File: packages/clay-navigation-bar/src/index.tsx [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+							{items.map(({active, href, label}, index) => (
+								<li
+									className={classNames('nav-item', {
+										active,
+									})}
+									data-nav-item-index={index}
+									key={index}
+									onClick={onItemClicked}
+								>
+									{href ? (
```
Comment (matuzalemsteles) [inline packages/clay-navigation-bar/src/index.tsx]: hey @bryceosterhaus and @diegonvs I've been wondering if we can use composition here too, as we need to pass `nav-item` and `active`, we can clone the elements, this would be simpler and people can easily customize and choose between link and button and add classes if necessary, for someone who wants to create their own component, we can send a warning stating that the `className` API is needed or we can offer more information about the item that was clicked with the `onItemClicked` method.

File: packages/clay-navigation-bar/src/index.tsx [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+						'navbar-toggler',
+						'navbar-toggler-link',
+						{
+							collapsed: !visible,
+						}
+					)}
+					displayType="secondary"
+					onClick={handleClickToggler}
+				>
+					{items.map(({active, label}) => {
```
Comment (matuzalemsteles) [inline packages/clay-navigation-bar/src/index.tsx]: It might be best to use `items.find(item => item.active)`

File: packages/clay-navigation-bar/package.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+	},
+	"main": "lib/index.js",
+	"module": "src/index.js",
+	"jsnext:main": "src/index.js",
+	"files": [
+		"lib",
+		"src"
+	],
+	"scripts": {
+		"test": "jest --config ../../jest.config.js"
```
Comment (matuzalemsteles) [inline packages/clay-navigation-bar/package.json]: We still need these scripts:

```
"build": "cross-env NODE_ENV=production babel src --root-mode upward --out-dir lib --extensions .ts,.tsx",
"build:types": "cross-env NODE_ENV=production tsc --project ./tsconfig.declarations.json",
"prepublishOnly": "npm run build && npm run build:types",
```

File: packages/clay-navigation-bar/webpack.config.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
+/**
```
Comment (matuzalemsteles) [inline packages/clay-navigation-bar/webpack.config.js]: You can remove this file, we do not need it anymore.

File: packages/clay-navigation-bar/stories/index.tsx [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+import ClayNavigationBar from '../src/index';
+import React from 'react';
+import {storiesOf} from '@storybook/react';
+
+import 'clay-css/lib/css/atlas.css';
+
+const spritemap = require('clay-css/lib/images/icons/icons.svg');
+
+storiesOf('ClayNavigationBar', module)
+	.add('default', () => (
```
Comment (matuzalemsteles) [inline packages/clay-navigation-bar/stories/index.tsx]: As we had discussed, you can already add the knobs in this PR to narrow down the list of examples. If the knob is not enough to show a particular use case, add a story.

File: packages/clay-navigation-bar/src/index.tsx (hunk 8 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/index.tsx]: Makes sense :)

**Comment by diegonvs** [comment]
@bryceosterhaus and @matuzalemsteles, PR updated :)

File: .vscode/launch.json [hunk 12]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (bryceosterhaus) [inline .vscode/launch.json]: We should remove this directory or add it to .gitignore

File: migrate-the-clay-components-from-v2-to-v3.md [hunk 13]
```diff
@@ -139,6 +139,45 @@ To get to the behavior of having a ClayLink with image, use the composition with
 +</ClayLink>
 ```
 
+## ClayNavigationBar
+
+ClayNavigationBar has become more simpler than the version `v2`, removing APIs from `items` and turning more flexible to pass elements by composition. Turning the development more easier when defining your Dropdown's list items but complying with Lexicon specifications.
```
Comment (bryceosterhaus) [inline migrate-the-clay-components-from-v2-to-v3.md]: Should be either "more simple" or "simpler"

File: packages/clay-navigation-bar/README.md [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+2. Install local dependencies:
+
+```
+yarn
+```
+
+3. Build the code:
+
+```
+yarn start
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/README.md]: We can remove this portion of the readme since its all through storybook now

File: packages/clay-navigation-bar/src/Helpers.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
+/**
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/Helpers.ts]: Since this file is not a component, we can leave it lowercase `helpers.ts`

File: packages/clay-navigation-bar/src/Hooks.ts [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+import {useEffect, useState} from 'react';
+
+export function useTransition(
+	visible: Boolean,
+	setVisible: any,
+	contentRef: React.RefObject<any>
+) {
+	const [transitioning, setTransitioning] = useState<Boolean>(false);
+
+	useEffect(() => {
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/Hooks.ts]: I wonder if this should be "useLayoutEffect." Any chance you happen to know when to use effect vs layoutEffect, i've been a bit confused by those two.

File: packages/clay-navigation-bar/src/Hooks.ts [hunk 17]
```diff
@@ -0,0 +1,55 @@
+/**
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/Hooks.ts]: This file should be lowercase since it isnt exporting a component by default. `hooks.ts`

File: packages/clay-navigation-bar/src/index.tsx [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+			{...otherProps}
+			className={classNames('nav-item', {
+				active,
+				className,
+			})}
+		>
+			{React.Children.map(
+				children,
+				(child: React.ReactElement<ItemProps>, index) => {
+					return React.cloneElement(child, {
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/index.tsx]: We can remove the return here and use an implicit return from the arrow function

File: packages/clay-navigation-bar/src/index.tsx [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+
+	const activeElementsCount = children.filter(child => child.props.active)
+		.length;
+
+	warning(
+		!(activeElementsCount > 1),
+		`You passed ${activeElementsCount} active props to ClayNavigationBar children.`
+	);
+
+	warning(
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/index.tsx]: Is this necessary since we require it via typescript params?

File: packages/clay-navigation-bar/src/index.tsx [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+		handleTransitionEnd,
+		handleClickToggler,
+	] = useTransition(visible, setVisible, contentRef);
+
+	const activeElementsCount = children.filter(child => child.props.active)
+		.length;
+
+	warning(
+		!(activeElementsCount > 1),
+		`You passed ${activeElementsCount} active props to ClayNavigationBar children.`
```
Comment (wincent) [inline packages/clay-navigation-bar/src/index.tsx]: This warning would be more helpful if it indicated *what* the expected count is. eg. something like:

```javascript
warning(
  activeElementsCount <= 1,
  `ClayNavigationBar expects 0 or 1 active children, but received ${activeElementsCount}`
);
```
`<tangential-open-question>`I also wonder whether we want to fail a bit harder in cases like this, or in general (ie. by throwing an `Error` when APIs are used incorrectly). [They say](https://en.wikipedia.org/wiki/Robustness_principle) that you should be "liberal in what you accept, strict in what you provide", but they [also say](https://en.wikipedia.org/wiki/Fail-fast) "fail fast". In most situations, I find the latter approach to lead to higher quality.`</tangential-open-question>`

File: .vscode/launch.json (hunk 12 above)
Comment (diegonvs) [reply .vscode/launch.json]: I miss this file, sorry 😝

File: packages/clay-navigation-bar/src/Hooks.ts (hunk 16 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/Hooks.ts]: Makes sense changing to useLayoutEffect @bryceosterhaus! `useEffect` can be used in the most of use cases but when you are mutating DOM elements we need use `useLayoutEffect`.

- https://kentcdodds.com/blog/useeffect-vs-uselayouteffect

File: packages/clay-navigation-bar/src/index.tsx (hunk 19 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/index.tsx]: Nope. Removing!

File: packages/clay-navigation-bar/src/index.tsx (hunk 20 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/index.tsx]: Makes sense! When creating this warning these principles wasn't in my mind. Thanks for guiding @wincent!

File: packages/clay-navigation-bar/src/Hooks.ts (hunk 16 above)
Comment (matuzalemsteles) [reply packages/clay-navigation-bar/src/Hooks.ts]: I'm not sure if this should be a `useLayoutEffect`, it will cause a new rendering and will only mutate after it completes the latest renderings... And it can cause a blocking visual updates (which in this case is what we are trying to do here), I'd rather go with `useEffect` if it causes problems we can go to `useLayoutEffect`.

File: migrate-the-clay-components-from-v2-to-v3.md [hunk 21]
```diff
@@ -139,6 +139,45 @@ To get to the behavior of having a ClayLink with image, use the composition with
 +</ClayLink>
 ```
 
+## ClayNavigationBar
+
+ClayNavigationBar has become simpler than the version `v2`, removing APIs from `items` and turning more flexible to pass elements by composition. Turning the development more easier when defining your Dropdown's list items but complying with Lexicon specifications.
```
Comment (matuzalemsteles) [inline migrate-the-clay-components-from-v2-to-v3.md]: I do not know if we would say that we are more in compliance with the Lexicon, since we are giving more flexibility, we are leaving slightly off the curve with the specifications.

File: packages/clay-navigation-bar/src/index.tsx [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+			className={classNames('nav-item', {
+				active,
+				className,
+			})}
+		>
+			{React.Children.map(
+				children,
+				(child: React.ReactElement<ItemProps>, index) =>
+					React.cloneElement(child, {
+						...child.props,
```
Comment (matuzalemsteles) [inline packages/clay-navigation-bar/src/index.tsx]: Since we are accepting anything within our Item, as a simple button, I am wondering if this can have problems with this and force using `child.props.className` too, you can check out these use cases.

File: package.json [hunk 23]
```diff
@@ -52,8 +52,8 @@
 		"lerna": "^3.4.0",
 		"prettier-eslint-cli": "^4.7.0",
 		"raf": "^3.4.1",
-		"react": "^16.8.1",
-		"react-dom": "^16.8.1",
+		"react": "^16.9.0-alpha.0",
```
Comment (bryceosterhaus) [inline package.json]: I don't think we should be committing any alpha versions, lets try to only rely on stable versions.

File: packages/clay-navigation-bar/src/__tests__/index.tsx [hunk 24]
```diff
@@ -0,0 +1,270 @@
...
+			</ClayNavigationBar>
+		);
+
+		expect(testRenderer.toJSON()).toMatchSnapshot();
+	});
+
+	it('renders a dropdown when clicking the collapsed element from NavigationBar', async () => {
+		const {container} = render(
+			<ClayNavigationBar
+				data-testid="#1"
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/__tests__/index.tsx]: Is this testid actually used? I didn't see anything relating to it.

File: packages/clay-navigation-bar/src/__tests__/index.tsx [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+
+		expect(navigationBarDropdown).toBeDefined();
+
+		expect(navigationBarDropdown).toMatchSnapshot();
+	});
+
+	it('collapses the dropdown expanded when trigger element is clicked', async () => {
+		const {container} = render(
+			<ClayNavigationBar
+				data-testid="#1"
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/__tests__/index.tsx]: ^

File: packages/clay-navigation-bar/src/hooks.ts [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+
+		if (!transitioning) {
+			setTransitioning(true);
+		}
+	};
+
+	return [transitioning, handleTransitionEnd, willTriggerTransition] as [
+		Boolean,
+		(event: React.TransitionEvent) => void,
+		(event: React.MouseEvent | MouseEvent) => void
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/hooks.ts]: Is this necessary to cast all of these types? I would've thought that TS would be able to infer it from the type of  the variable.

File: packages/clay-navigation-bar/src/index.tsx [hunk 27]
```diff
@@ -0,0 +1,161 @@
...
+
+type ItemType = React.FunctionComponent<ItemProps>;
+
+const Item: ItemType = ({active, children, className, ...otherProps}) => {
+	return (
+		<li
+			{...otherProps}
+			className={classNames('nav-item', {
+				active,
+				className,
```
Comment (bryceosterhaus) [inline packages/clay-navigation-bar/src/index.tsx]: This will render `className` literally as a class if they exist. I think this needs to be moved outside of the object.

File: package.json (hunk 23 above)
Comment (diegonvs) [reply package.json]: This version is the only one that supports putting asynchronous logic on `act(` of react-testing-library.

Comment (diegonvs) [reply package.json]: https://github.com/facebook/react/releases/tag/v16.9.0-alpha.0

Comment (bryceosterhaus) [reply package.json]: Since this is only for a test case, can we just disable to test case until a stable version is released?

File: migrate-the-clay-components-from-v2-to-v3.md [hunk 28]
```diff
@@ -139,6 +139,45 @@ To get to the behavior of having a ClayLink with image, use the composition with
...
+     inverted
+     spritemap={spritemap}
+ >
++    <ClayNavigationBar.Item active>
++        <ClayLink className="nav-link" displayType="secondary" href="#1">
++            <span className="navbar-text-truncate">Page 1</span>
++        </ClayLink>
++    </ClayNavigationBar.Item>
++
++    <ClayNavigationBar.Item handle>
```
[the remark this answers is not in the document]
Comment (diegonvs) [reply migrate-the-clay-components-from-v2-to-v3.md]: Miss typo :(

File: packages/clay-navigation-bar/src/__tests__/index.tsx (hunk 24 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/__tests__/index.tsx]: Nope, thanks for finding

File: packages/clay-navigation-bar/src/hooks.ts (hunk 26 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/hooks.ts]: Removing these types Typescript isn't infering the type of these parameters. See the following screenshot:

Screen Shot 2019-05-09 at 18 40 14

File: packages/clay-navigation-bar/src/index.tsx (hunk 27 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/index.tsx]: yep!

File: package.json (hunk 23 above)
Comment (diegonvs) [reply package.json]: Makes sense. However our coverage will be decreased until the next React release.

File: packages/clay-navigation-bar/src/__tests__/index.tsx (hunk 25 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/__tests__/index.tsx]: Removing...

File: packages/clay-navigation-bar/src/index.tsx (hunk 22 above)
Comment (diegonvs) [reply packages/clay-navigation-bar/src/index.tsx]: Makes sense Matu! I'm thinking about passing `active` via context instead the className.

Comment (diegonvs) [reply packages/clay-navigation-bar/src/index.tsx]: I added some examples using HTMLElements as child of `<ClayNavigationBar.Item>` and everything worked fine.

**Comment by diegonvs** [comment]
Hey @matuzalemsteles, @bryceosterhaus: PR Updated!

**Comment by diegonvs** [comment]
Resent at https://github.com/liferay/clay/pull/1978
a
